vm/ubpf_vm.c

```diff
@@ -1188,31 +1198,53 @@ bounds_check(
 {
     if (!vm->bounds_check_enabled)
         return true;
-    if (mem && (addr >= mem && ((char*)addr + size) <= ((char*)mem + mem_len))) {
-        /* Context access */
-        return true;
-    } else if (addr >= stack && ((char*)addr + size) <= ((char*)stack + UBPF_STACK_SIZE)) {
-        /* Stack access */
-        return true;
-    } else if (
-        vm->bounds_check_function != NULL &&
-        vm->bounds_check_function(vm->bounds_check_user_data, (uintptr_t)addr, size)) {
-        /* Registered region */
-        return true;
-    } else {
+
+    uintptr_t access_start= (uintptr_t)addr;
+    uintptr_t access_end = access_start + size;
+    uintptr_t stack_start = (uintptr_t)stack;
+    uintptr_t stack_end = stack_start + UBPF_STACK_SIZE;
+    uintptr_t mem_start = (uintptr_t)mem;
+    uintptr_t mem_end = mem_start + mem_len;
+
+    if (access_start > access_end) {
         vm->error_printf(
             stderr,
-            "uBPF error: out of bounds memory %s at PC %u, addr %p, size %d\nmem %p/%zd stack %p/%d\n",
+            "uBPF error: invalid memory access %s at PC %u, addr %p, size %d\n",
             type,
             cur_pc,
             addr,
-            size,
-            mem,
-            mem_len,
-            stack,
-            UBPF_STACK_SIZE);
+            size);
         return false;
     }
+
+    // Check if the access is within the memory bounds.
+    if (access_start >= mem_start && access_end <= mem_end) {
+        return true;
+    }
+
+    // Check if the access is within the stack bounds.
+    if (access_start >= stack_start && access_end <= stack_end) {
+        return true;
+    }
+
+    // Check if the access is within the bounds of a registered region.
+    if (vm->bounds_check_function != NULL && vm->bounds_check_function(vm->bounds_check_user_data, access_start, size)) {
```

Comment:
How will this work if there is no bounds check function? It seems like we would get an error in that case. Is that the expected semantics?

Reply:
I think that makes perfect sense. Because validation will only happen for member about which it is unaware, if there is no bounds checking function, it *is* an error? If I am understanding correctly, then we are all good!